xmr-btc/src/lib.rs

```diff
@@ -50,3 +50,151 @@ pub mod bitcoin;
+
+            let tx_redeem = bitcoin::TxRedeem::new(&tx_lock, &redeem_address);
+            let tx_redeem_encsig = b.encsign(S_a_bitcoin.clone(), tx_redeem.digest());
+
+            co.yield_(Action::SendBitcoinRedeemEncsig(tx_redeem_encsig.clone()))
+                .await;
+
+            let tx_redeem_published = bitcoin_ledger
+                .watch_for_raw_transaction(tx_redeem.txid())
+                .await
+                .expect("TODO: implementor of this trait must make it infallible by retrying");
+
+            // NOTE: If any of this fails, Bob will never be able to take the monero.
+            // Therefore, there is no way to handle these errors other than aborting
+            let tx_redeem_sig = tx_redeem
+                .extract_signature_by_key(tx_redeem_published, b.public())
+                .expect("redeem transaction must include signature from us");
+            let s_a = bitcoin::recover(S_a_bitcoin, tx_redeem_sig, tx_redeem_encsig).expect(
+                "alice can only produce our signature by decrypting our encrypted signature",
+            );
+            let s_a = monero::PrivateKey::from_scalar(monero::Scalar::from_bytes_mod_order(
+                s_a.to_bytes(),
+            ));
+
+            let s_b = monero::PrivateKey {
+                scalar: s_b.into_ed25519(),
+            };
+
+            co.yield_(Action::CreateMoneroWalletForOutput {
+                spend_key: s_a + s_b,
+                view_key: v,
+            })
+            .await;
+
+            Ok(())
+        };
+
+        // NOTE: swap result should only be `Err` if we have reached the
+        // `refund_timelock`. Therefore, we should always yield the refund action
+        if swap_result.is_err() {
+            let tx_cancel =
+                bitcoin::TxCancel::new(&tx_lock, refund_timelock, A.clone(), b.public());
+            let tx_refund = bitcoin::TxRefund::new(&tx_cancel, &refund_address);
+
+            let signed_tx_cancel = {
+                let sig_a = tx_cancel_sig_a.clone();
+                let sig_b = b.sign(tx_cancel.digest());
+
+                tx_cancel
+                    .clone()
+                    .add_signatures(&tx_lock, (A.clone(), sig_a), (b.public(), sig_b))
+                    .expect("sig_{a,b} to be valid signatures for tx_cancel")
+            };
+
+            let signed_tx_refund = {
+                let adaptor = Adaptor::<Sha256, Deterministic<Sha256>>::default();
+
+                let sig_a =
+                    adaptor.decrypt_signature(&s_b.into_secp256k1(), tx_refund_encsig.clone());
+                let sig_b = b.sign(tx_refund.digest());
+
+                tx_refund
+                    .add_signatures(&tx_cancel, (A.clone(), sig_a), (b.public(), sig_b))
+                    .expect("sig_{a,b} to be valid signatures for tx_refund")
+            };
+
+            co.yield_(Action::RefundBitcoin {
```

Comment:
Ideally, our swap execution state machines need to get even more fine granular than that where after yielding such an action, we watch for it to appear in the mempool, monitor if the fees we set are appropriate and if not, emit another action that bumps the fees until the TX gets included in time.

Reply:
That is a cool idea with the fee bumping! I had a similar thought when I considered splitting this action into two, and only yield the second one if the first has sufficient confirmations. This means that the caller doesn't need to consider all this logic.